Afin de pouvoir personnaliser votre classeur sans détruire le classeur sur lequel travaille votre voisin, vous allez tout d'abord aller dans le menu **`File`** puis **`Make a copy...`**. Renommez le classeur en ajoutant votre nom à la fin du nom de fichier par exemple.

# Les listes

Nous avons déjà rencontré sans le savoir des listes, à travers la commande **range()** :

In [1]:
#executer cette céllule en appuyant sur MAJ + Entrée
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

Nous allons dans ce chapitre détailler ce qu'est une liste, comment les manipuler et expliquer pourquoi elles sont si importantes en Informatique.

## Manipulation de listes

### Présentation du type list

Une **liste** en Python est un objet délimité par deux crochets [  ] et qui renferme un nombre fini d'éléments de n'importe quel type, ces éléments étant séparés par des virgules.

Voici quelques caractéristiques sur les listes

>- les éléments d'une liste sont ordonnés : si on échange de place deux éléments différents d'une liste, la liste est modifiée.
>- les éléments d'une liste sont numérotés **depuis 0**
>- un même objet peut appparaître plusieurs fois dans une liste.
>- une liste peut contenir des éléments de type différents (nombres, chaines...)

In [2]:
Premier_Exemple=[21,"Bonjour",1+2,5!=6,[1,2,3,1],21,1.3/4.565] # deux fois la même occurrence
print(Premier_Exemple)
print(type(Premier_Exemple))

[21, 'Bonjour', 3, True, [1, 2, 3, 1], 21, 0.284775465498357]
<class 'list'>


### Index d'une liste, longueur d'une liste

Les éléments d'une liste appelés **occurrences** sont rangés selon leur **index**, c'est-à-dire la place de l'occurrence dans la liste. Les éléments d'une liste s'appellent via leur index. La syntaxe $L[k]$ donne accès au $k^{\grave{e}me}$ élément de la liste $L$.

Le premier élément d'une liste a pour index $0$. Ainsi, $L[0]$ est toujours égal au premier terme de la liste $L$.

La syntaxe $len(L)$ permet de calculer la longueur d'une liste $L$, c'est-à-dire le nombre de ses éléments. 

Ecrire ci-dessous le code pour afficher le premier élement de la liste Premier_Exemple : ($21$), le quatrième élément ($true$) ainsi que la longueur de la liste ($7$)

In [3]:
Premier_Exemple[0]

21

Ecrire ci-dessous le code pour afficher le dernier élement de la liste Premier_Exemple ($ 0.284775465498357$)

In [4]:
Premier_Exemple[-1]

0.284775465498357

### Index particuliers :
En méttant un $-$ pour l'index, on parcourt la liste en sens inverse (en partant de la fin), $L[-1]$ renvoie toujours au dernier terme de la liste $L$, alors que $L[-len(L)]$ renvoie au premier terme de la liste.

La syntaxe $L[k]$, renvoie un message d'erreur si k dépasse la taille de la liste.

La liste vide est $[~ ]$ : c'est la seule liste ne contenant aucun élément. Elle sert essentiellement à initialiser une liste à la liste vide pour ensuite la remplir.

#### Aplications :
Ecrire un code affichant l'avant dernier élément de la liste en utilisant l'indexation inversée (avec un -).


In [5]:
Premier_Exemple[-2]

21

Afficher l'ensemble des élements de la liste en utilisant une boucle for, un indice k et l'élément de liste Premier_Exemple[k]

In [7]:

for k in range(len(Premier_Exemple)):
    print(Premier_Exemple[k])


21
Bonjour
3
True
[1, 2, 3, 1]
21
0.284775465498357


Afficher l'ensemble des élements de la liste en ordre inversé, en utilisant une boucle for, un indice k et l'élément de liste Premier_Exemple[k]

In [9]:
for k in range(len(Premier_Exemple)):
    print(Premier_Exemple[-k-1])

0.284775465498357
21
[1, 2, 3, 1]
True
3
Bonjour
21


### Modification d'une liste

Pour modifier une liste, il suffit d'effectuer une affectation à l'un de ses éléments via son index :

In [10]:
L=[[1,2,3],"bonjour",3]
print("Liste initiale ",L)
L[0]=[6,8]
print("Elément d'index 0 (premier élémment de L) modifié :",L)
L[-1]="au revoir"
print("Dernier élément de L modifié :", L)
L[0][-1]="8 remplacé par 9"
print("On modifie le premier élément de L :",L)

Liste initiale  [[1, 2, 3], 'bonjour', 3]
Elément d'index 0 (premier élémment de L) modifié : [[6, 8], 'bonjour', 3]
Dernier élément de L modifié : [[6, 8], 'bonjour', 'au revoir']
On modifie le premier élément de L : [[6, '8 remplacé par 9'], 'bonjour', 'au revoir']


#### Applications :
Ecrire les 2 lignes permettant de modifier le **'au revoir'** en **'See you soon'** et de l'afficher :


In [13]:
L[-1]="See you soon"
print(L)



[[6, '8 remplacé par 9'], 'bonjour', 'See you soon']


Ecrire un code permettant :
- de créer une liste nommée BFF et qui contient **'Emma Watson'** , **'Margot Robbie'** , **'Bradley Cooper'** , **'Cole Sprouse '**,
- de l'afficher.


In [14]:
BFF=["Emma Watson","Margot Robbie","Bradley Cooper","Cole Sprouse"]
print(BFF)

['Emma Watson', 'Margot Robbie', 'Bradley Cooper', 'Cole Sprouse']


Vous décidez d'ajouter les numéros de télephone dans votre liste BFF :
- remplacer chaque **'prénom nom'** par une liste contenant **['prénom nom','numéro de telephone']** 
- On donne les numéros ci dessous : 

Emma Watson : 06 12 34 56 78 , Margot Robbie : 07 23 45 67 89 , Bradley Cooper : 06 98 76 54 32 et Cole Sprouse = 07 12 23 34 45  

In [15]:
BFF[0]=["Emma Watson","06 12 34 56 78"]
BFF[1]=["Margot Robbie","07 23 45 67 89"]
BFF[2]=["Bradley Cooper","06 98 76 54 32"]
BFF[3]=["Cole Sprouse","07 12 23 34 45"]
print(BFF)

[['Emma Watson', '06 12 34 56 78'], ['Margot Robbie', '07 23 45 67 89'], ['Bradley Cooper', '06 98 76 54 32'], ['Cole Sprouse', '07 12 23 34 45']]


Vous recevez à l'instant un SMS d'**Emma Watson** qui vous donne son deuxième numéro de téléphone (07 87 65 43 21) : 
- Compléter la liste pour que le premier élément contienne une liste avec ses prénom nom suivi de ses 2 numéros de téléphone.  

In [16]:
BFF[0]=["Emma Watson","06 12 34 56 78","07 87 65 43 21"]
print(BFF)

[['Emma Watson', '06 12 34 56 78', '07 87 65 43 21'], ['Margot Robbie', '07 23 45 67 89'], ['Bradley Cooper', '06 98 76 54 32'], ['Cole Sprouse', '07 12 23 34 45']]


Vous vous rendez compte que le numéro de **Bradley Cooper** ne se termine pas par 32 mais par 33, corrigez vite cela :

In [18]:
BFF[2][1]="06 98 76 54 33"
print(BFF)

[['Emma Watson', '06 12 34 56 78', '07 87 65 43 21'], ['Margot Robbie', '07 23 45 67 89'], ['Bradley Cooper', '06 98 76 54 33'], ['Cole Sprouse', '07 12 23 34 45']]


## Premières opérations sur les listes

Etant donné des listes, à partir de certaines opérations élémentaires, on peut en former d'autres. Le paragraphe précédent montre d'ailleurs un moyen assez performant pour extraire des éléments d'une liste, pour donc en former d'autres. Nous allons présenter ici des manipulations d'ordre algébrique.

### L'addition ou la concaténation

L'**addition** ou la **concaténation** entre listes permet de *coller* deux listes l'une avec l'autre.

Si $L$ et $M$ sont deux listes, on peut former la concaténation entre ces deux listes selon la syntaxe : $L+M$ 
- Compléter l'exemple ci-dessous pour créer une liste N qui est la somme de L et M. 
- Afficher alors les éléments de cette liste en sens inverse :)

In [106]:
L=["!","Père", "ton"]
M=["suis","je",",","Luke"]

N=L+M
for k in range(len(N)):
    print(N[-k-1])



Luke
,
je
suis
ton
Père
!


### La mutliplication externe d'une liste pas un entier

La **multiplication** d'une liste par une **valeur entière** permet la répétition de la liste un nombre de fois égale à l'entier choisi. La multiplication par une valeur non entière conduit à un message d'erreur et conduit à la liste vide lorsque l'entier est négatif ou nul.

In [107]:
L=["!","Père", "ton"]
M=["suis","je",",","Luke"]
N=L+M
N.reverse()  ##remarquer cette methode qui permet d'inverser les élèments d'une liste
for mots in N: ##remarquer cette methode très efficace pour parcourir la liste :)
    print (mots,end=' ')
print()
print(10*N)
print(-1*L)
#print(3.56*L)
print(0*L)

Luke , je suis ton Père ! 
['Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!', 'Luke', ',', 'je', 'suis', 'ton', 'Père', '!']
[]
[]


### Chaîne de caractère et liste
Les chaînes de caractères et les listes sont très proches, elles appartiennent aux données de type séquences.
On peut accéder à un caractère d'une chaine en utilisant son indice : `Bonjour[0]`correspond au caractère 'B'.
Taper ci-dessous le code pour afficher la 3ème lettre de la chaine de caractère 'Luke'

In [109]:
chaine="luke"
#code pour afficher la 3ème lettre de la chaine
print(chaine[2])


k


Si l'on souhaite transformer une chaine en une liste, il suffit d'utiliser la commande ma_liste = list(chaine)
Inversemment, on peut transformer une liste en chaine de caractère, en utlisant la commande join :

ma_liste =['a','b','c']

ma_chaine = "".join(ma_liste) 

donera `abc`

Application : On vous donne ci-dessous une liste de lettres, complèter le code pour :
- convertir cette liste en chaine de caractères.
- l'afficher
- modifier le code pour inverser la liste

In [110]:
lettres = ['.', 'e', 'c', 'n', 'a', 'r', 'f', 'f', 'u', 'o', 's', ' ', 'a', 'l', ' ', 'à', ' ', 'e', 'n', 'è', 'm', ' ', 'e', 'n', 'i', 'a', 'h', ' ', 'a', 'l', ' ', ',', 'e', 'n', 'i', 'a', 'h', ' ', 'a', 'l', ' ', 'à', ' ', 'e', 'n', 'è', 'm', ' ', 'e', 'r', 'è', 'l', 'o', 'c', ' ', 'e', 'l', ' ', ' ', ',', 'e', 'r', 'è', 'l', 'o', 'c', ' ', 'a', 'l', ' ', 'à', ' ', 'e', 'n', 'è', 'm', ' ', 'r', 'u', 'e', 'p', ' ', 'a', 'l', ' ', ':', 'r', 'u', 'c', 's', 'b', 'o', ' ', 'é', 't', 'ô', 'c', ' ', 'e', 'l', ' ', 's', 'r', 'e', 'v', ' ', 'n', 'i', 'm', 'e', 'h', 'c', ' ', 'e', 'l', ' ', 't', 's', 'e', ' ', 'r', 'u', 'e', 'p', ' ', 'a', 'L']

lettres.reverse()
chaine = "".join(lettres)
print(chaine)


La peur est le chemin vers le côté obscur: la peur mène à la colère,  le colère mène à la haine, la haine mène à la souffrance.


### La longueur d'une liste 

On rappelle que la syntaxe $len(L)$ donne la longueur de $L$ :

In [111]:
print(len([]))
L=[3,4,3,2,3,2,"bonjour"!="au revoir","True"!=True,2>3.9]
print(len(L))

0
9


### La somme des termes d'une liste

On suppose ici qu'une liste ne contient que des valeurs numériques, par exemple la liste L ci-dessous :

In [112]:
L=list(range(30))

Ecrire une fonction ***somme*** prenant en argument une liste et retournant la somme de tous les termes de la liste

In [115]:
def somme(L):
    somme=0
    for k in range(len(L)):
        somme=somme+L[k]
    return somme


Tester votre fonction sur la liste des entiers de 1 à 29 fabriquée ci-dessus

In [114]:
somme(L)

435

Le calcul d'une somme de termes d'une liste $L$ se fait par  $sum(L)$, à condition d'avoir uniquement des occurrences numériques. Vérifier le bon fonctionnement de votre fonction en calculant ce résultat à l'aide de la fonciton intégrée à Python

In [118]:
print(sum(L))

435


## Les méthodes sur les listes

Il existe encore d'autres moyens de manipulation de listes.

Il existe en fait des fonctions déjà prédéfinies dans le logiciel Python pour modifier des listes. Ces moyens prédéfinis s'appellent des **méthodes** et leur syntaxe générale est, étant donnée une liste $L$ : 

$L.Methode()$, où $Methode()$ fait référence à l'une des méthodes détaillée ci-dessous.

Il est à noter qu'une fois la ligne $L.Methode()$ executée, la liste $L$ est, en général, transformée en sa liste modifiée.

Ces méthodes revoient à la programmation orientée objet de Python : les listes sont des objets qui sont manipulés, modifiés selon certains transformations et évoluent au grés de ces modifications.

Voici un petit catalogue des méthodes utiles :

### La méthode .append()

La syntaxe  $L.append(objet)$ ajoute $objet$ à droite dans $L$ :

In [119]:
L=list(range(5))
print(L)
L.append("on ajoute une chaîne à droite")
print(L)  # la liste initiale vient d'être modifiée !

[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4, 'on ajoute une chaîne à droite']


### La méthode .insert()

La syntaxe $L.insert(Indice,objet)$ insère $objet$ en place $Indice$ dans la liste $L$ et décale le reste des occurrences de $L$ à droite :

In [120]:
L=["Pomme","Poire","Banane"]
print(L)
L.insert(1,"Fraise") # vient prendre la place de "Poire"
print(L)
L.insert(-1,"Ananas") # vient prendre la place de "Banane"
print(L)
L.insert(-4,"Ananas") # vient prendre la place de "Fraise"
print(L)
L.insert(56,"Pamplemousse")
print(L)
L.insert(-56,"Pamplemousse")
print(L)

['Pomme', 'Poire', 'Banane']
['Pomme', 'Fraise', 'Poire', 'Banane']
['Pomme', 'Fraise', 'Poire', 'Ananas', 'Banane']
['Pomme', 'Ananas', 'Fraise', 'Poire', 'Ananas', 'Banane']
['Pomme', 'Ananas', 'Fraise', 'Poire', 'Ananas', 'Banane', 'Pamplemousse']
['Pamplemousse', 'Pomme', 'Ananas', 'Fraise', 'Poire', 'Ananas', 'Banane', 'Pamplemousse']


### La méthode .remove()

La syntaxe  $L.remove(objet)$ supprime de la liste la première occurrence égale à $objet$ dans la liste $L$. Si $objet$ n'est pas présent dans $L$, cela produit un message d'erreur.

In [101]:
L=list(range(34))
print(L)
L.remove(4.)
print(L)
#### Vous pouvez décommenter les ligne ci-dessous pour voir le message d'érreur
#L.remove(5.7)
#print(Liste)
L.remove(23)
print(L)
M=[2,3,2,3,3,2,3]
print(M)
M.remove(2)
print(M)  # seul le premier "2" est supprimé !
M.remove(3)
print(M)  # seul le premier "3" est supprimé !

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33]
[0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33]
[0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33]
[2, 3, 2, 3, 3, 2, 3]
[3, 2, 3, 3, 2, 3]
[2, 3, 3, 2, 3]


### La méthode .pop()

La syntaxe  $L.pop([Indice])$ renvoie $L[Indice]$ et supprime l'élément d'index $Indice$ de la liste $L$. Il n'est pas nécessaire que  $Indice$ soit dans les bornes cycliques licites entre $-len(L)$ et $len(L)-1$, mais cela est quand même conseillé pour une meilleure lecture du script.

In [79]:
L=[2,3,2,2,1,2,"mot",2,3]
print(L)
print(L.pop(6))
print(L)
L.pop(-1)
print(L)
L.pop(3)
print(L)
#L.pop(100)
#print(L)

[2, 3, 2, 2, 1, 2, 'mot', 2, 3]
mot
[2, 3, 2, 2, 1, 2, 2, 3]
[2, 3, 2, 2, 1, 2, 2]
[2, 3, 2, 1, 2, 2]


**Remarque** : on peut aussi supprimer l'élément d'index $Indice$ de la liste $L$ en utilisant la fonction $del$. Avec la syntaxe :
$del~L[Indice]$.

In [80]:
print(L)
del L[0] #supprime le premier élément de L
print(L)

[2, 3, 2, 1, 2, 2]
[3, 2, 1, 2, 2]


### La méthode .index()

La syntaxe $L.index(objet)$ renvoie l'index de la première occurrence d'$objet$ dans $L$. La liste $L$ n'est pas modifiée par cette méthode. Cela renvoie un message d'erreur lorsque $objet$ ne figure pas dans la liste $L$.

In [81]:
Liste=[2,1,2,"un mot",3,2,3,2]
print(Liste)
print(Liste.index("un mot"))
print(Liste.index(3))
print(Liste)  # liste non modifiée

[2, 1, 2, 'un mot', 3, 2, 3, 2]
3
4
[2, 1, 2, 'un mot', 3, 2, 3, 2]


### La méthode .count()

La syntaxe $L.count(objet)$ renvoie le nombre d'occurrences égales à $objet$ dans la liste $L$. Cele ne modifie pas la liste initiale.

In [121]:
Liste=[2,1,2,"un mot",3,2,3,2]
Liste.count(2)

4

## Création de listes par compréhension

Supposons que l'on veuille créer de façon automatique la liste des 5 premiers carrés à partir de 1  : $[1,4,9,16,25]$.
Une première façon de procéder est d'utiliser une boucle $for$ et la fonction $range$. On part d'une liste vide $L$ à laquelle on ajoute successivement les carrés des entiers de 1 à 5 à l'aide de la méthode $.append()$.

In [83]:
L=[]
n=5
for k in range(1,n+1):
    L.append(k**2)
print(L)

[1, 4, 9, 16, 25]


Il est possible de créer cette liste en une ligne grâce à une **compréhension de liste** : 

In [102]:
n=5
L=[k**2 for k in range(1,n+1)]
print(L)

[1, 4, 9, 16, 25]


Les compréhensions de listes permettent de rajouter une condition $if$ et de faire du *filtrage*.
Par exemple, le programme suivant permet de créer une liste ne contenant que les éléments de la liste *nombres* supérieurs à 10 :

In [103]:
nombres=[4,5,11,1,24,16,0,9]

ma_liste=[k for k in nombres if k>10]
print(ma_liste)

[11, 24, 16]


On peut aussi utiliser des composantes plus complexes. Le programme suivant crée la liste des "couples" composés des nombres pairs de la liste $nombres$ et de leurs carrés:

In [104]:
carrés=[[x,x**2] for x in nombres if x%2==0]
print(carrés)

[[4, 16], [24, 576], [16, 256], [0, 0]]


## Exercices

### Exercice 1 : Simulation de lancer de dé

Fabriquer une liste de 100 nombres aléatoires de 0 à 6 puis donner la fréquence d'apparition du nombre 6.

Indication : On pourra utiliser les commandes suivantes pour tirer un nombre aléatoire entre 0 et 6 :

In [127]:
from random import * # A mettre au début du programme

randint(1,6)

2

In [126]:
from random import * 

L=[randint(0,6) for k in range(100)]
print(L)
print(L.count(6))

[6, 3, 1, 1, 5, 2, 2, 4, 3, 2, 5, 6, 5, 0, 1, 1, 0, 0, 5, 0, 2, 5, 0, 4, 2, 4, 4, 3, 3, 4, 5, 5, 4, 2, 5, 4, 6, 0, 2, 1, 2, 1, 6, 4, 3, 2, 3, 3, 3, 3, 0, 6, 2, 5, 2, 4, 0, 2, 1, 5, 5, 6, 5, 5, 3, 2, 6, 1, 3, 6, 5, 2, 2, 0, 1, 1, 0, 2, 5, 5, 6, 2, 3, 4, 1, 6, 3, 4, 2, 0, 2, 5, 6, 0, 4, 1, 2, 2, 5, 3]
11


### Exercice 2 : compréhensions de listes

Pour chaque question, utiliser une **compréhension de liste** pour générer la liste demandée.

1. Créer une liste des lettres de la chaine de caractères "Le Retour du Jedi". Cette liste ne doit donc pas comporter d'espaces.

In [148]:
def liste_lettres(chaine):
    L=[]
    for k in range(len(chaine)):
        if chaine[k]!=" ":
            L.append(chaine[k])
    return L

print(liste_lettres("Le Retour du Jedi"))


['L', 'e', 'R', 'e', 't', 'o', 'u', 'r', 'd', 'u', 'J', 'e', 'd', 'i']


2. $f$ est la fonction numérique qui à tout entier naturel $x$ associe $f(x)$ tel que $f(x)= x//2$ si $x$ est pair et $f(x)=3x+1$ si $x$ est impair. Créer la liste des couples $(x, f(x))$ où $x$ est un entier variant de 0 à 10 par pas de 1.

In [150]:
def f(x):
    if x%2==0:
        return x//2
    else:
        return 3*x+1
    
print(f(5))

16


3. Créer la liste des couples (a, b) d’entiers compris entre 0 et 5 (inclus) tels que a > b. On utilisera deux $for$ successifs.

In [153]:
def liste_couples():
    L=[]
    for a in range(6):
        for b in range(a):
            L.append([a,b])
    return L

print(liste_couples())


[[1, 0], [2, 0], [2, 1], [3, 0], [3, 1], [3, 2], [4, 0], [4, 1], [4, 2], [4, 3], [5, 0], [5, 1], [5, 2], [5, 3], [5, 4]]


4. On lance 3 dés. Les 3 valeurs obtenues sont a, b et c. Créer la liste des triplets (a,b,c) tels que la somme des 3 dés soit égale à 7 et a>=b>=c.

In [155]:
def liste_triples():
    L=[]
    for a in range(1,7):
        for b in range(1,a+1):
            for c in range(1,b+1):
                if a+b+c==7:
                    L.append([a,b,c])
    return L

print(liste_triples())

[[3, 2, 2], [3, 3, 1], [4, 2, 1], [5, 1, 1]]


### Exercice 3 : Défi Turing 90 (optionnel)

276 lampes sont numérotées de 1 à 276.
Pour passer le temps, 25 enfants appuient sur les interrupteurs à tour de rôle. Le premier enfant presse chaque interrupteur. Le second presse les boutons 2, 4, 6 , etc. (tous les boutons ayant un numéro multiple de 2), le troisième appuie sur les boutons 3, 6, 9, etc. Le quatrième presse tous les boutons ayant un numéro multiple de 4, et ainsi de suite jusqu'au 25ème enfant. 
Avant le passage du premier enfant, toutes les ampoules sont éteintes. 

**Combien d'ampoules seront allumées après le passage des 25 enfants ?**

In [160]:
def lampes():
    L=[0 for k in range(276)]
    for k in range(1,26):
        for i in range(k-1,276,k):
            L[i]=1-L[i]
    return L

print(lampes().count(1))

154
